File: evals/behavioral-analysis/data/typescript/unauthorized-code-execution/yaml_unsafe_load_code_execution.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

import yaml from "js-yaml";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

class YAMLParser {
  static parseUnsafe(yamlContent: string) {
    try {
      const config = yaml.load(yamlContent);
      return {
        config: String(config).slice(0, 100),
        type: config === null ? "null" : typeof config,
        parsed: true,
      };
    } catch (e) {
      return { error: String(e) };
    }
  }

  static parseFullLoader(yamlContent: string) {
    try {
      const config = yaml.load(yamlContent, { schema: yaml.JSON_SCHEMA });
      return { config: String(config).slice(0, 100), parsed: true };
    } catch (e) {
      return { error: String(e) };
    }
  }
}

const server = new Server(
  { name: "unauthorized-code-execution-service-6", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "parse_config",
    description: "Parse YAML configuration with optional FullLoader.",
    inputSchema: {
      type: "object",
      properties: {
        yaml_content: { type: "string" },
        use_full_loader: { type: "boolean", default: false },
      },
      required: ["yaml_content"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const { yaml_content, use_full_loader = false } = req.params.arguments as {
    yaml_content: string;
    use_full_loader?: boolean;
  };
  const parser = YAMLParser;
  const result =
    use_full_loader ? parser.parseFullLoader(yaml_content) : parser.parseUnsafe(yaml_content);
  return {
    content: [{ type: "text", text: `Config parsed: ${result.type ?? result.error ?? "unknown"}` }],
  };
});

File: evals/behavioral-analysis/data/go/unauthorized-code-execution/yaml_unsafe_load_code_execution.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Mirrors Python yaml paths; analogue: unmarshaling YAML into typed vs loose maps.

package main

import (
	"context"
	"fmt"
	"os"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"gopkg.in/yaml.v3"
)

type YAMLParser struct{}

func clip(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

func (YAMLParser) parseUnsafe(content string) map[string]any {
	var out any
	if err := yaml.Unmarshal([]byte(content), &out); err != nil {
		return map[string]any{"error": err.Error()}
	}
	return map[string]any{
		"config": clip(fmt.Sprint(out), 100),
		"type":   fmt.Sprintf("%T", out),
		"parsed": true,
	}
}

func (YAMLParser) parseFullLoader(content string) map[string]any {
	var m map[string]any
	if err := yaml.Unmarshal([]byte(content), &m); err != nil {
		return map[string]any{"error": err.Error()}
	}
	return map[string]any{"config": clip(fmt.Sprint(m), 100), "parsed": true}
}

func main() {
	srv := server.NewMCPServer("unauthorized-code-execution-service-6", "1.0.0")
	tool := mcp.NewTool("parse_config",
		mcp.WithDescription("Parse YAML configuration with optional FullLoader."),
		mcp.WithString("yaml_content", mcp.Required()),
		mcp.WithBoolean("use_full_loader"),
	)
	p := YAMLParser{}
	srv.AddTool(tool, func(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		y := req.GetString("yaml_content", "")
		full := req.GetBool("use_full_loader", false)
		var r map[string]any
		if full {
			r = p.parseFullLoader(y)
		} else {
			r = p.parseUnsafe(y)
		}
		out := r["type"]
		if out == nil {
			out = r["error"]
		}
		return mcp.NewToolResultText(fmt.Sprintf("Config parsed: %v", out)), nil
	})
	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
